File: Rust/src/release.rs
//repo - Handles users github public projects
use crate::assets::Asset;
#[allow(non_snake_case)]

//TODO: Implement Date Parse

pub struct Release { Name: String, Tag: String, URL: String, Info: String, Created: String, Published: String, isPreRelease: bool, parseCorrectly: bool, Assets: Vec<Asset>}

impl Release
{
    pub fn get_name(&self) -> &String { &self.Name }
    pub fn get_tag(&self) -> &String { &self.Tag}
    pub fn get_url(&self) -> &String { &self.URL}
    pub fn get_info(&self) -> &String { &self.Info }
    pub fn get_created_date(&self) -> &String { &self.Created}
    pub fn get_published_date(&self) -> &String { &self.Published }
    pub fn get_isprerelease(&self) -> &bool { &self.isPreRelease }
    pub fn get_parsecorrectly(&self) -> &bool { &self.parseCorrectly}
    pub fn get_assets(&self) -> &Vec<Asset> { &self.Assets }

    pub(crate) fn AnyUnassigned(arr: [String; 7]) -> bool
    {
        //https://www.programming-idioms.org/idiom/110/check-if-string-is-blank/1761/rust

        //Loop through the array from 0 to 6 (7 Elements, n-1 = 6)
        for x in  0..6 { if arr[x].trim().is_empty() == true { return false } }

        //Return true as block in loop hasn't been called
        return true
    }

    //The "fake" constructor needs an Array of "String" in size of "7"
    pub(crate) fn new(Result: [String; 7]) -> Self
    {
        Self
        {
            /*
                Array should be 7/8 in length
                Result[0] = Name ("name")
                Result[1] = Tag ("tag_name")
                Result[2] = URL ("html_url")
                Result[3] = INFO ("body")
                Result[4] = Create Date ("created_at")
                Result[5] = Published Date ("published_at")
                Result[6] = PreRelease ? true/false ("prerelease")
            */

            Name: Result[0].clone(),
            Tag: Result[1].clone(),
            Info: Result[2].clone(),
            URL: Result[3].clone(),
            Created: Result[4].clone(),
            Published: Result[5].clone(),

            //Return true if Result[6] is equal to "true" or "True", any else - false.
            isPreRelease: match Result[6].as_str() { "true" => {true} "True" => {true} _ => {false}},
            parseCorrectly: Repo::AnyUnassigned(Result),
            Assets: Vec::new()
        }
    }
}

File: Python/Classes/Release.py
# Release.py - Handles the release(s) information

import datetime  # Import time module for "__parse_date"
from Assets import Asset  # Import Assets.py into Release.py


class Release:

    def __parse_date(self, _input: str):
        # https://stackabuse.com/converting-strings-to-datetime-in-python/
        date_time_obj = datetime.datetime.strptime(_input, "%Y-%m-%dT%H:%M:%SZ")
        return f"{date_time_obj.date()} {date_time_obj.time()}"

    def __init__(self, result: list):

        # First check if array size is 7
        if not len(result) == 7:
            raise Exception(
                f" \"result\" in Release.__init__ should be of size \"7\", but got {len(result)} instead...")

        '''
            Array should be 7 / 8 in length
            Result[0] = Name("name")
            Result[1] = Tag("tag_name")
            Result[2] = URL("html_url")
            Result[3] = INFO("body")
            Result[4] = Create Date("created_at")
            Result[5] = Published Date("published_at")
            Result[6] = PreRelease ? true / false("prerelease")
        '''

        try:
            self._Name = result[0]
            self._Tag = result[1]
            self._URL = result[2]
            self._Info = result[3]
            self._Created = self.__parse_date(result[4])
            self._Published = self.__parse_date(result[5])
            self._isPreRelease = True if result[6] == "true" or result[6] == "true" else False
            self._parsedCorrectly = True
        except (ValueError, Exception):  # <- PyCharm throws "too broad exception" if it doesn't include this
            self._parsedCorrectly = False
        self._Assets = []  # Create new list (empty for now)

    def add_asset(self, assets: Asset):
        if self._Assets.__contains__(assets):
            print(f"[INFO] Ignoring {assets.item_name()} as it is already included into the list")
        else:
            self._Assets.append(assets)

    def get_assets(self):
        return self._Assets

    def get_name(self):
        return self._Name

    def get_tag(self):
        return self._Tag

    def get_url(self):
        return self._URL

    def get_description(self):
        return self._Info

    def get_created_date(self):
        return self._Created

    def get_published_date(self):
        return self._Published

    def get_is_prerelease(self) -> bool:
        return self._isPreRelease

    def get_is_parsedcorrectly(self) -> bool:
        return self._parsedCorrectly


''' Testing goes here

test_data = [
    "Beta Build 0.4.2",
    "0.4.2",
    "https://github.com/TheE7Player/CSGO-Event-Viewer/releases/tag/0.4.2",
    "# Beta build 0.4.2 ... ...",
    "2019-12-08T15:39:26Z",
    "2019-12-08T15:44:45Z",
    "false"
]

e = Release(test_data)

print(f"STATUS: {e.get_is_parsedcorrectly()}")
print(f"Name: {e.get_name()}")
print(f"Tag: {e.get_tag()}")
print(f"Url: {e.get_url()}")
print(f"Info: {e.get_description()}")
print(f"Date Created: {e.get_created_date()}")
print(f"Date Public: {e.get_published_date()}")
print(f"Prerelease: {e.get_is_prerelease()}")

i1 = Asset('Item_1', 'URL HAHAHA', 2951456)
i2 = Asset('Item_2', 'URL HAHAHA', 2951456)
e.add_asset(i1)
e.add_asset(i2)
nList = e.get_assets()
'''
